rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+
+#### Establishing Initial Checkpoint
+
+The writer first establishes it’s checkpoint when starting up, as part of incrementing the epoch:
+1. Read the latest manifest version V.
+2. Compute the writer’s epoch E
+3. Look through the checkpoints under `checkpoint` for any entries with `metadata` of type `WriterCheckpoint`, and remove any such entries.
+4. Create a new checkpoint with id set to a new v4 uuid, `manifest_id` set to the next manifest version (V + 1), and `checkpoint_expire_time_s` set to None, and `metadata` set to `WriterCheckpoint{ epoch: E }`.
+5. Set `writer_epoch` to E
+6. Write the manifest. If this fails with a CAS error, go back to 1.
+
+#### Refreshing the Checkpoint
+
+The writer will periodically re-establish its checkpoint to pin the latest SSTs and SRs. It will run the following whenever it detects that the manifest has been updated. This happens either when the manifest is polled (at `manifest_poll_interval`), or when the writer detects an update when it goes to update the manifest with it’s local view of the core db state:
+1. Read the latest manifest version V. (If the writer is applying an update from the local state then in the first iteration it can assume that it’s local version is the latest. If it’s wrong, the CAS will fail and it should reload the manifest from Object Storage))
+2. Decide whether it needs to re-establish it’s checkpoint. The writer reestablishes the checkpoint whenever the set of L0 SSTs, or Sorted Runs has changed. This includes both changes made locally by the Writer, or by the Compactor.
+    1. If so, create a new checkpoint with `manifest_id` set to V+1, `checkpoint_expire_time` set to None, and `metadata` set to `WriterCheckpoint{ epoch: E }`.
+    2. Delete the old checkpoint from `checkpoints`. When we support range scans, we’ll instead want to maintain references to the checkpoint from the range scans and delete the checkpoint when all references are dropped. (We should technically do this for point lookups too, but the likelihood of GC deleting an SST in this case is very low).
+    3. Apply any local changes
+    4. Write the new manifest. If this fails with a CAS error, go back to 1.
+
+#### Clones
+
+The writer will also support initializing a new database from an existing checkpoint. This allows users to “fork” an instance of slatedb, allowing it to access all the original db’s data from the checkpoint, but isolate writes to a new db instance.
+
+To support this, we add the `Db::open_from_checkpoint` method, which accepts a `source_path` and `source_checkpoint`. When initializing the db for the first time (detected by the presence of a manifest), the writer will do the following:
```

Comment:
isn't it by *absence* of a manifest instead of *presence*?

Reply:
Yep, I'll clarify.